sensor/common/detector/baseline/baseline.go

```diff
@@ -75,16 +89,199 @@ func (w *baselineEvaluator) AddBaseline(baseline *storage.ProcessBaseline) {
+	for _, contentHash := range containerMap {
+		oe.removeReference(contentHash)
+	}
+
+	delete(oe.deploymentBaselines, id)
+	log.Debugf("Successfully removed deployment %s from baseline", id)
+}
+
+// AddBaseline adds a process baseline to the optimized evaluator
+func (oe *optimizedBaselineEvaluator) AddBaseline(baseline *storage.ProcessBaseline) {
+	oe.lock.Lock()
+	defer oe.lock.Unlock()
+
+	deploymentID := baseline.GetKey().GetDeploymentId()
+	containerName := baseline.GetKey().GetContainerName()
+
+	// Check if baseline should be unlocked (has UserLockedTimestamp = nil)
+	if baseline.GetUserLockedTimestamp() == nil {
+		oe.removeBaseline(deploymentID, containerName, baseline)
+		return
+	}
+
+	// Locked baseline - process normally
+	baselineSet := set.NewStringSet()
+	for _, elem := range baseline.GetElements() {
+		if process := elem.GetElement().GetProcessName(); process != "" {
+			baselineSet.Add(process)
+		}
+	}
+
+	// Find existing process set with same content or create new one
+	contentHash := oe.findOrCreateProcessSet(baselineSet)
+
+	// Update deployment mapping
+	if oe.deploymentBaselines[deploymentID] == nil {
+		oe.deploymentBaselines[deploymentID] = make(map[string]string)
+	}
+
+	// If this deployment/container already has a process set, decrement its ref count
+	if oldContentHash, exists := oe.deploymentBaselines[deploymentID][containerName]; exists {
+		oe.removeReference(oldContentHash)
+	}
+
+	oe.deploymentBaselines[deploymentID][containerName] = contentHash
+	log.Debugf("Successfully added locked process baseline %s", baseline.GetId())
+}
+
+func (oe *optimizedBaselineEvaluator) removeBaseline(deploymentID string, containerName string, baseline *storage.ProcessBaseline) {
+	log.Debugf("Removing (id:%s, UserLockedTimestamp:%v, elements:%v)", baseline.GetId(), baseline.GetUserLockedTimestamp(), baseline.GetElements())
+	if oe.deploymentBaselines[deploymentID] != nil {
+		if oldContentHash, exists := oe.deploymentBaselines[deploymentID][containerName]; exists {
+			oe.removeReference(oldContentHash)
+			delete(oe.deploymentBaselines[deploymentID], containerName)
+			if len(oe.deploymentBaselines[deploymentID]) == 0 {
+				delete(oe.deploymentBaselines, deploymentID)
+			}
+		} else {
+			log.Debugf("Baseline for container name %s does not exist", containerName)
+		}
+	} else {
+		log.Debugf("Baseline for deployment ID %s does not exist", deploymentID)
+	}
+}
+
+// findOrCreateProcessSet finds an existing process set with the same content or creates a new one
+func (oe *optimizedBaselineEvaluator) findOrCreateProcessSet(processes set.StringSet) string {
+	contentHash := computeProcessSetHash(processes)
+
+	// Check if we already have this process set
+	if entry, exists := oe.processSets[contentHash]; exists {
+		// Check for hash collision and verify content actually matches
+		if !entry.processes.Equal(processes) {
+			log.Panic("SHA256 hash collision detected for process set %v vs existing %v",
```

Comment:
Blocker: I commented on that already in another comment, but it looks like we are still panicking here. Lets use a wrapper like `utils.Should(error)` which would panic in dev, but only error in prod.

Reply:
Fixed in https://github.com/stackrox/stackrox/pull/17019